src/Bicep.Core/Analyzers/Linter/Rules/NoHardcodedLocationRule.cs

```diff
@@ -0,0 +1,199 @@
+        // should use a parameter, an expression (but not resourceGroup().location or
+        // deployment().location) or the string 'global'.
+
+        public new const string Code = "no-hardcoded-location";
+
+        public NoHardcodedLocationRule() : base(
+            code: Code,
+            description: CoreResources.NoHardcodedLocationRuleDescription,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}"))
+        {
+        }
+
+        public override string FormatMessage(params object[] values)
+            => string.Format((string)values[0]);
+
+        public override IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model)
+        {
+            RuleVisitor visitor = new RuleVisitor(this, model);
+            visitor.Visit(model.SourceFile.ProgramSyntax);
+            return visitor.diagnostics;
+        }
+
+        /// <summary>
+        /// Find a name starting with `baseName` that is not already used as a top-level definition
+        /// </summary>
+        private static string GetUnusedTopLevelName(string baseName, SemanticModel model)
+        {
+            int increment = 1;
+            while (true)
+            {
+                string newName = increment == 1 ? baseName : $"{baseName}{increment}";
+                if (!model.Root.GetDeclarationsByName(newName).Any())
+                {
+                    return newName;
+                }
+
+                ++increment;
+            }
+        }
+
+        private void ValidateResourceLocationValue(List<IDiagnostic> diagnostics, SyntaxBase locationValueSyntax, SemanticModel model, string? moduleParameterName)
+        {
+            // Is the value a string literal (or a variable defined as a string literal)?
+            (string? literalValue, VariableSymbol? definingVariable) = TryGetLiteralTextValueAndDefiningVariable(locationValueSyntax, model);
+            if (literalValue == null)
+            {
+                // No - it's okay
+                return;
+            }
+
+            if (StringComparer.OrdinalIgnoreCase.Equals(literalValue, Global))
+            {
+                // The string value 'global' (case-insensitive) is allowed
+                return;
+            }
+
+            if (definingVariable != null)
+            {
+                // It's using a variable that is defined as a literal string. Suggest they change it to
+                // a parameter (the error goes on the variable definition, not the resource location property)
+                //
+                // e.g.
+                //
+                // var location = 'westus'    << suggest change this to param
+                // resource ... {
+                //   location: location
+
+                TextSpan errorSpan = definingVariable.NameSyntax.Span;
+
+                // Is there already a diagnostic for this variable definition? Don't add a duplicate
+                if (diagnostics.Any(d => d.Code == Code && d.Span.Equals(errorSpan)))
```

Comment:
>Any [](http://example.com/codeflow?start=32&length=3)

Is a list scan here ok or do we need to optimize the time complexity?

Reply:
I'd be extremely surprised, but since it is n^2 went ahead and did it.